### PR TITLE
Add `tsh kubectl` support for tracer exporter

### PR DESCRIPTION
**Pull request by tigrato**
This PR adds support for `--trace` exporter when running `tsh kubectl`.
It allows any user to automatically export traces from `tsh`.

Usage: `tsh --trace --trace-exporter=dir kubectl get pods`

### CODE REVIEW

File: tool/tsh/kubectl.go [hunk 1]
```diff
@@ -101,17 +105,63 @@ func runKubectlReexec(selfExec string, args []string, collector io.Writer) error
...
+			// only update the provider if we successfully set it up
+			cf.TracingProvider = provider
+
+			// ensure that the provider is shutdown on exit to flush any spans
+			// that haven't been forwarded yet.
+			closeTracer = func() {
+				shutdownCtx, cancel := context.WithTimeout(cf.Context, 1*time.Second)
+				defer cancel()
+				err := provider.Shutdown(shutdownCtx)
+				if err != nil && !strings.Contains(err.Error(), context.DeadlineExceeded.Error()) {
```
Comment (codingllama) [inline tool/tsh/kubectl.go:146]: Could we use errors.Is here instead?

File: tool/tsh/kubectl.go [hunk 2]
```diff
@@ -101,17 +105,63 @@ func runKubectlReexec(selfExec string, args []string, collector io.Writer) error
...
 // This code is only executed when `tshKubectlReexec` env is present. This happens
 // because we need to retry kubectl calls and `kubectl` calls os.Exit in multiple
 // paths.
-func runKubectlCode(args []string) {
+func runKubectlCode(cf *CLIConf, args []string) {
+	closeTracer := func() {}
+	if cf.SampleTraces {
+		provider, err := newTraceProvider(cf, "", nil)
+		if err != nil {
+			log.WithError(err).Debug("failed to set up span forwarding.")
```
Comment (codingllama) [inline tool/tsh/kubectl.go:135]: nit: Capitalize log message:

```suggestion
			log.WithError(err).Debug("Failed to set up span forwarding")
```

File: tool/tsh/kubectl.go [hunk 3]
```diff
@@ -101,17 +105,63 @@ func runKubectlReexec(selfExec string, args []string, collector io.Writer) error
...
+				defer cancel()
+				err := provider.Shutdown(shutdownCtx)
+				if err != nil && !strings.Contains(err.Error(), context.DeadlineExceeded.Error()) {
+					log.WithError(err).Debugf("failed to shutdown trace provider")
+				}
+			}
+		}
+	}
+
+	ctx, span := cf.TracingProvider.Tracer("kubectl").Start(cf.Context, "kubectl")
```
Comment (codingllama) [inline tool/tsh/kubectl.go:153]: Is this safe when `cf.SampleTraces` is not set? Maybe leave a comment explaining why it's alright to always use `cf.TracingProvider`.

Comment (tigrato) [reply tool/tsh/kubectl.go:153]: It's safe. cf.TracingProvider is pre-initialized with a noop provider.

File: tool/tsh/kubectl.go (hunk 1 above)
Comment (tigrato) [reply tool/tsh/kubectl.go:146]: done

**Review by codingllama — APPROVED** [review]

**Review by gzdunek — APPROVED** [review]

**Comment by greedy52** [comment]
@tigrato may I backport this to v13? I want to backport my other PR but seems to depend on this. This PR doesn't seem to be "risky" either.

**Pull request merged**